splice_machine/src/main/java/com/splicemachine/derby/impl/storage/SpliceRegionAdmin.java

```diff
@@ -0,0 +1,408 @@
+        ExecRow execRow = getExecRow(td, index);
+        if (execRow != null) {
+            DataHash dataHash = getEncoder(td, index, execRow);
+            KeyHashDecoder decoder =  dataHash.getDecoder();
+            decoder.set(rowKey, 0, rowKey.length);
+            decoder.decode(execRow);
+        }
+
+        ArrayList<ExecRow> rows=new ArrayList<>(1);
+        DataValueDescriptor[] dvds=new DataValueDescriptor[]{
+                new SQLVarchar(),
+        };
+
+        ExecRow row=new ValueRow(dvds.length);
+        row.setRowArray(dvds);
+        dvds[0] = new SQLChar(execRow != null ? execRow.toString() : Bytes.toStringBinary(rowKey));
+
+        rows.add(row);
+        ResultColumnDescriptor[] columnInfo=new ResultColumnDescriptor[1];
+        columnInfo[0]=new GenericColumnDescriptor("START_KEY",
+                DataTypeDescriptor.getBuiltInDataTypeDescriptor(Types.VARCHAR,50));
+
+        EmbedConnection defaultConn=(EmbedConnection) SpliceAdmin.getDefaultConn();
+        Activation lastActivation=defaultConn.getLanguageConnection().getLastActivation();
+        IteratorNoPutResultSet resultsToWrap=new IteratorNoPutResultSet(rows,columnInfo,lastActivation);
+        resultsToWrap.openCore();
+        results[0] = new EmbedResultSet40(defaultConn,resultsToWrap,false,null,true);
+
+    }
+
+    /**
+     *
+     * @param rowKey rowkey for primary key or index value
+     * @param conglomId conglomerate number for splice table
+     * @return a region that contains the provided rowkey
+     * @throws IOException
+     */
+    private static Partition getPartition(byte[] rowKey, long conglomId) throws IOException {
+
+        PartitionAdmin admin= SIDriver.driver().getTableFactory().getAdmin();
+        Iterable<? extends Partition> partitions =  admin.allPartitions(Long.toString(conglomId));
+        List<Partition> partitionList = Lists.newArrayList(partitions);
+        partitionList.sort(new Comparator<Partition>() {
+            @Override
+            public int compare(Partition o1, Partition o2) {
+                byte[] k1 = o1.getStartKey();
+                byte[] k2 = o2.getStartKey();
+                if (k1 == null || k1.length == 0) {
+                    if (k2 == null || k2.length == 0)
+                        return 0;
+                    else
+                        return -1;
+                }
+                else {
+                    if (k2 == null || k2.length == 0)
+                        return 1;
+                    else
+                        return Bytes.compareTo(k1, k2);
+                }
+            }
+        });
+
+        //create a list of start keys
+        byte[][] startKeys = new byte[partitionList.size()][];
+        for (int i = 0; i < partitionList.size(); ++i) {
+            Partition p = partitionList.get(i);
+            byte[] k = p.getStartKey();
+            startKeys[i] = k;
+        }
+        int index = binarySearchStartKey(startKeys, rowKey);
```

Comment:
It seems to me here we are doing more than we should. At this point, we've already conducted a sort (NlogN), then a full list scan (N) to prepare the startKeys list.  Then the binary search (logN). so complexity is dominant by the sort.
But We could have just got the region that we need by a simple scan of the partition list with complexity of N.